src/common/core/coretree.js

```diff
@@ -38,55 +37,74 @@ define([
         this.loadPaths = TASYNC.wrap(storage.loadPaths);
         this.logger = logger;
 
+        function ASSERT_IS_OBJECT(value) {
+            ASSERT(value !== null && typeof value === 'object' && value instanceof Array === false);
+        }
+
         // ------- memory management
 
         function __detachChildren(node) {
-            ASSERT(node.children instanceof Array && node.age >= CONSTANTS.MAX_AGE - 1);
+            ASSERT_IS_OBJECT(node.children);
+            ASSERT(node.age >= CONSTANTS.MAX_AGE - 1);
 
             var children = node.children;
             node.children = null;
             node.age = CONSTANTS.MAX_AGE;
 
-            for (var i = 0; i < children.length; ++i) {
-                __detachChildren(children[i]);
+            for (var child in children) {
+                __detachChildren(children[child]);
             }
         }
 
         function __ageNodes(nodes) {
-            ASSERT(nodes instanceof Array);
+            ASSERT_IS_OBJECT(nodes);
 
-            var i = nodes.length;
-            while (--i >= 0) {
-                var node = nodes[i];
+            var toDelete = [],
+                node,
+                i,
+                key;
 
+            for (key in nodes) {
+                node = nodes[key];
                 ASSERT(node.age < CONSTANTS.MAX_AGE);
                 if (++node.age >= CONSTANTS.MAX_AGE) {
-                    nodes.splice(i, 1);
+                    toDelete.push(key);
                     __detachChildren(node);
                 } else {
                     __ageNodes(node.children);
                 }
             }
+
+            for (i = 0; i < toDelete.length; i += 1) {
+                delete nodes[toDelete[i]];
+            }
         }
 
         function __ageRoots() {
+            var root,
+                i;
             if (++ticks >= CONSTANTS.MAX_TICKS) {
                 ticks = 0;
-                __ageNodes(roots);
+                for (i = 0; i < roots.length; i += 1) {
+                    root = roots[i];
+                    ASSERT(root.age < CONSTANTS.MAX_AGE);
+                    if (++root.age >= CONSTANTS.MAX_AGE) {
+                        roots.splice(i, 1);
```

Comment:
if you want to splice inside the loop, then you should use decreasing iterator not to skip any elements